pkg/server/plugin/notifier/k8sbundle/k8sbundle_test.go

```diff
@@ -0,0 +1,391 @@
+	s.raw.hooks.newKubeClient = func(configPath string) (kubeClient, error) {
+		s.Equal(expectedConfigPath, configPath)
+		if client == nil {
+			return nil, errors.New("kube client not configured")
+		}
+		return client, nil
+	}
+}
+
+func (s *Suite) configure(configuration string) {
+	_, err := s.p.Configure(context.Background(), &spi.ConfigureRequest{
+		Configuration: configuration,
+	})
+	s.Require().NoError(err)
+}
+
+type fakeIdentityProvider struct {
+	mu      sync.Mutex
+	bundles []*common.Bundle
+}
+
+func newFakeIdentityProvider() *fakeIdentityProvider {
+	return &fakeIdentityProvider{}
+}
+
+func (c *fakeIdentityProvider) FetchX509Identity(ctx context.Context, req *hostservices.FetchX509IdentityRequest) (*hostservices.FetchX509IdentityResponse, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if len(c.bundles) == 0 {
+		return nil, errors.New("no bundle")
+	}
+	bundle := c.bundles[0]
+	c.bundles = c.bundles[1:]
+	// Send back the bundle. Leave the Identity field empty, since it is
+	// unused by the notifier.
+	return &hostservices.FetchX509IdentityResponse{
+		Bundle: bundle,
+	}, nil
+}
+
+func (c *fakeIdentityProvider) appendBundle(bundle *common.Bundle) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.bundles = append(c.bundles, bundle)
+}
+
+type fakeKubeClient struct {
+	mu         sync.Mutex
+	configMaps map[string]*corev1.ConfigMap
+	patchErr   error
+}
+
+func newFakeKubeClient(configMaps ...*corev1.ConfigMap) *fakeKubeClient {
+	c := &fakeKubeClient{
+		configMaps: make(map[string]*corev1.ConfigMap),
+	}
+	for _, configMap := range configMaps {
+		c.setConfigMap(configMap)
+	}
+	return c
+}
+
+func (c *fakeKubeClient) GetConfigMap(ctx context.Context, namespace, configMap string) (*corev1.ConfigMap, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	entry, ok := c.configMaps[configMapKey(namespace, configMap)]
+	if !ok {
+		return nil, errors.New("not found")
+	}
+	return entry, nil
+}
+
+func (c *fakeKubeClient) PatchConfigMap(ctx context.Context, namespace, configMap string, patchBytes []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	entry, ok := c.configMaps[configMapKey(namespace, configMap)]
+	if !ok {
+		return errors.New("not found")
+	}
+
+	// if there is a patch error configured, return it and clear the patchErr state.
+	patchErr := c.patchErr
+	c.patchErr = nil
+	if patchErr != nil {
+		return patchErr
+	}
+
+	patchedMap := new(corev1.ConfigMap)
+	if err := json.Unmarshal(patchBytes, patchedMap); err != nil {
+		return err
+	}
+	resourceVersion, err := strconv.Atoi(patchedMap.ResourceVersion)
+	if err != nil {
+		return errors.New("patch does not have resource version!")
+	}
+	entry.ResourceVersion = fmt.Sprint(resourceVersion + 1)
+	if entry.Data == nil {
+		entry.Data = map[string]string{}
+	}
+	for key, data := range patchedMap.Data {
+		entry.Data[key] = data
+	}
+	return nil
+}
+
+func (c *fakeKubeClient) getConfigMap(namespace, configMap string) *corev1.ConfigMap {
```

Comment:
is this function really needed? Could GetConfigMap just be used?

Reply:
it was convenient to not have to deal with the error when checking the config map contents in tests (which will fail gracefully if nil is returned). I've changed the implementation of GetConfigMap to use getConfigMap, if that helps.